bigframes/core/transpiler/googlesql/query.py

```diff
@@ -0,0 +1,151 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import dataclasses
+import typing
+
+import bigframes.core.transpiler.googlesql.abc as abc
+import bigframes.core.transpiler.googlesql.expression as expr
+
+"""This module provides a structured representation of GoogleSQL syntax using nodes.
+Each node's name and child nodes are designed to strictly follow the official GoogleSQL
+syntax rules outlined in the documentation:
+https://cloud.google.com/bigquery/docs/reference/standard-sql/query-syntax"""
+
+
+@dataclasses.dataclass
+class QueryExpr(abc.SQLSyntax):
+    """This class represents GoogleSQL `query_expr` syntax."""
+
+    select: Select
+    with_cte_list: typing.Sequence[NonRecursiveCTE] = ()
+
+    def sql(self) -> str:
+        text = []
+        if len(self.with_cte_list) > 0:
+            with_cte_text = ",\n".join(
+                [with_cte.sql() for with_cte in self.with_cte_list]
+            )
+            text.append(f"WITH {with_cte_text}")
+
+        text.append(self.select.sql())
+        return "\n".join(text)
+
+
+@dataclasses.dataclass
+class Select(abc.SQLSyntax):
+    """This class represents GoogleSQL `select` syntax."""
+
+    select_list: typing.Sequence[SelectExpression]
+    from_clause_list: typing.Sequence[FromClause] = ()
+
+    def sql(self) -> str:
+        text = ["SELECT"]
+
+        select_list_sql = ",\n".join([select.sql() for select in self.select_list])
+        text.append(select_list_sql)
+
+        if self.from_clause_list is not None:
+            from_clauses_sql = ",\n".join(
+                [clause.sql() for clause in self.from_clause_list]
+            )
+            text.append(f"FROM\n{from_clauses_sql}")
+        return "\n".join(text)
+
+
+@dataclasses.dataclass
+class SelectExpression(abc.SQLSyntax):
+    """This class represents `select_expression` and `select_all` (aka. `SELECT *`)
```

Comment:
Might be worth having two node types so we don't need the custom validation logic anymore

Reply:
Done. Yeah, we can choose to follow googlesql syntxt restrictly.